server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java

```diff
@@ -1318,7 +1360,21 @@ public static IndexMetaData fromXContent(XContentParser parser) throws IOExcepti
                                 throw new IllegalArgumentException("Unexpected token: " + token);
                             }
                         }
-                    } else if ("warmers".equals(currentFieldName)) {
+                    } else if (KEY_PRIMARY_TERMS.equals(currentFieldName)) {
```

Comment:
why did you add deserialization code here?

Is it because we are testing API deserialization somewhere? If so, I would rather have the tests changed to test non-API deserialization.

Same comment of other places below.

Reply:
The primary reason for adding these is, the original `ClusterState.toXContent` emits array serialization for `aliases` and object serialization for `primary_terms`, which are different from what `MetaData.toXContent` emits, which is object serialization for `aliases` and array serialization for `primary_terms`, under API context. The delta can been seen here https://github.com/elastic/elasticsearch/pull/52743/commits/072e05d370f464ec8582d55c7c26bb8199b58735#diff-4385d650a8cef0592db3454ad42e4bf3L1172-R1157 

In general, as the duplicated code diverges, there are conflicting serializations under API context. 

So the additional deserialization here is to be able to read back in the different serialization emitted by the new code (namely, array serialization for `aliases` and object serialization for `primary_terms`)

Reply:
I think we should avoid being able to read back the API-serialized XContent. There is no requirement to do so, and it introduces leniency in the parsing here, with the possibility of silently hiding bugs where the wrong context is chosen for serialization.

Reply:
I've given this a shot and removed deserialization for both `primary_terms` and `aliases`. These ended up breaking a lot of tests as they serialized under default context (API), and deserialized for further checking. I thus removed the breaking tests (as little as I can, but more can be removed as what's left are just setup code) and isolated the removal in this commit  https://github.com/elastic/elasticsearch/pull/52743/commits/0c11613501e979d9202a1f6fa7dda3aca4b69b72. 

If we don't want to remove these many tests for such small changes, the other possible approach is to not emit those different serialization for `aliases` and `primary_terms` for `ClusterState.toXContent`. This obviously would break code relying on it.

Please let me know which way you prefer, or if you have other ideas around this.